### PR TITLE
feat(coinmarket): Trade box

### LINKED ISSUE #9359: Trade box in account
### Design
Design WIP - TBA

### Spec 

**Issue**
Users are unsure of the meaning of the Trade button and unaware of the Trading options in Suite.

**Solution**
image
image

In addition to the Trade button in header we will show a Trade box in each account with relevant coin, showing 7d change in value.

### PR DESCRIPTION
**Pull request by FreeWall**
## Description

Trade box in account

## Related Issue

Resolve #9359, #9711

## Screenshots:
image

### CODE REVIEW

File: suite-common/wallet-types/src/fiatRates.ts
Comment (dahaca) [inline suite-common/wallet-types/src/fiatRates.ts]: Just `WeeklyRates` and no need to have a duplicate?

File: packages/suite-analytics/src/constants.ts
Comment (dahaca) [inline packages/suite-analytics/src/constants.ts:39]: nit: 
```suggestion
    AccountsTradeboxAction= 'accounts/tradebox',
```

File: suite-common/wallet-core/src/fiat-rates/fiatRatesMiddleware.ts
Comment (dahaca) [inline suite-common/wallet-core/src/fiat-rates/fiatRatesMiddleware.ts]: Unwanted fetch in mobile, at least for now.

File: suite-common/wallet-core/src/fiat-rates/fiatRatesThunks.ts
Comment (dahaca) [inline suite-common/wallet-core/src/fiat-rates/fiatRatesThunks.ts]:
```suggestion
    `${actionPrefix}/updateLastTwoWeekRates`,
```

File: suite-common/wallet-core/src/fiat-rates/fiatRatesThunks.ts
Comment (dahaca) [inline suite-common/wallet-core/src/fiat-rates/fiatRatesThunks.ts]: How about having just 1 action that takes the amount of weeks as an argument? `blockbookService.fetchLastWeekRates` is just a wrapper for `getFiatRatesForTimestamps.getMultiTickers` anyways :) Not a fan of having all those duplicate actions and types.

File: packages/suite/src/views/wallet/transactions/TradeBox/TradeBox.tsx
Comment (dahaca) [inline packages/suite/src/views/wallet/transactions/TradeBox/TradeBox.tsx:108]: nit: create a selector, also to be used in `Ticker`

File: packages/suite/src/views/wallet/transactions/TradeBox/TradeBox.tsx
Comment (dahaca) [inline packages/suite/src/views/wallet/transactions/TradeBox/TradeBox.tsx]:
```suggestion
    const getRateAge = (timestamp: number) => differenceInMinutes(new Date(timestamp), new Date());
```

File: packages/suite/src/views/wallet/transactions/TradeBox/TradeBox.tsx
Comment (dahaca) [inline packages/suite/src/views/wallet/transactions/TradeBox/TradeBox.tsx]:
```suggestion
    ${variables.SCREEN_QUERY.MOBILE} {
```

File: packages/suite/src/views/wallet/transactions/TradeBox/TradeBox.tsx
Comment (dahaca) [inline packages/suite/src/views/wallet/transactions/TradeBox/TradeBox.tsx]: `Footer`?

File: packages/suite/src/components/suite/OverflowItems.tsx
Comment (dahaca) [inline packages/suite/src/components/suite/OverflowItems.tsx]: this behavior exists in [AppNavigation](https://github.com/trezor/trezor-suite/blob/89834958042c4e046a0d8faf4e316b3ca37066e7/packages/suite/src/components/suite/AppNavigation.tsx#L211) as well

Comment (dahaca) [inline packages/suite/src/components/suite/OverflowItems.tsx]:
```suggestion
export function OverflowItems<T extends Item>({
        items,
        wrapper: Wrapper = 'div',
        minVisibleItems = 0,
        alwaysRenderOverflow = false,
        visibleItemRenderer,
        overflowRenderer,
    }: OverflowItemsProps<T>) {
```

File: packages/suite/src/components/suite/OverflowItems.tsx
Comment (dahaca) [inline packages/suite/src/components/suite/OverflowItems.tsx]: So `growing` means that it's reset to all items being visible?

File: packages/suite/src/components/suite/OverflowItems.tsx
Comment (dahaca) [inline packages/suite/src/components/suite/OverflowItems.tsx]:
```suggestion
                setState({
                    overflowItems: [],
                    visibleItems: props.items,
                })
```

File: packages/suite/src/components/suite/OverflowItems.tsx
Comment (dahaca) [inline packages/suite/src/components/suite/OverflowItems.tsx]: nit: define the function outside of `setState` for readability.

File: packages/suite/src/components/suite/OverflowItems.tsx
Comment (dahaca) [inline packages/suite/src/components/suite/OverflowItems.tsx]: is there a reason why you are using a callback here?

File: packages/suite/src/components/suite/OverflowItems.tsx
Comment (dahaca) [inline packages/suite/src/components/suite/OverflowItems.tsx]: Could you explain what those do?

File: packages/suite/src/components/suite/OverflowItems.tsx
Comment (dahaca) [inline packages/suite/src/components/suite/OverflowItems.tsx]: Better do it with a styled component, just pass a className :)

File: packages/suite/src/components/suite/OverflowItems.tsx
Comment (dahaca) [inline packages/suite/src/components/suite/OverflowItems.tsx]: create a styled component

File: packages/suite/src/components/suite/OverflowItems.tsx
Comment (dahaca) [inline packages/suite/src/components/suite/OverflowItems.tsx:20]: I believe it would be better to avoid this renderer pattern since I believe this OverflowItems will always have an array of buttons which are folded into a dropdown. So you can just define [those](https://github.com/trezor/trezor-suite/blob/2726578bb5e95c8b96c3cef603deb511d1f15e6b/packages/suite/src/views/wallet/transactions/TradeBox/TradeBoxMenu.tsx#L93) right here, in `OverflowItems` and add an `onClick` and `data-test` fields to the `Item`

File: packages/suite/src/views/wallet/transactions/TradeBox/TradeBoxMenu.tsx
Comment (dahaca) [inline packages/suite/src/views/wallet/transactions/TradeBox/TradeBoxMenu.tsx]: `width: 100%` looks nice on those buttons, at least on mobile and tablet ;)

File: packages/suite/src/views/wallet/transactions/TradeBox/TradeBoxMenu.tsx
Comment (dahaca) [inline packages/suite/src/views/wallet/transactions/TradeBox/TradeBoxMenu.tsx]: use the `Item` type?

File: packages/suite/src/views/wallet/transactions/TradeBox/TradeBoxMenu.tsx
Comment (dahaca) [inline packages/suite/src/views/wallet/transactions/TradeBox/TradeBoxMenu.tsx]: is it needed? `Translation` already renders a `span`

File: packages/suite/src/views/wallet/transactions/TradeBox/TradeBoxMenu.tsx
Comment (dahaca) [inline packages/suite/src/views/wallet/transactions/TradeBox/TradeBoxMenu.tsx]: Why use custom paddings and fonts here?

Comment (dahaca) [reply packages/suite/src/views/wallet/transactions/TradeBox/TradeBoxMenu.tsx]: I see, it's copied from `AppNavigation`. I'd only leave the `font-weight` here and even that is controversial 😅

File: packages/suite/src/views/wallet/transactions/TradeBox/TradeBoxGraph.tsx
Comment (dahaca) [inline packages/suite/src/views/wallet/transactions/TradeBox/TradeBoxGraph.tsx]: use the one from `styled-components` please, the one in `@trezor/components` is gonna be deprecated soon :)

File: packages/suite/src/views/wallet/transactions/TradeBox/TradeBoxPrices.tsx
Comment (dahaca) [inline packages/suite/src/views/wallet/transactions/TradeBox/TradeBoxPrices.tsx]: nit: we nearly always put the component state at the top

File: packages/suite/src/views/wallet/transactions/TradeBox/TradeBoxPrices.tsx
Comment (dahaca) [inline packages/suite/src/views/wallet/transactions/TradeBox/TradeBoxPrices.tsx]: just use `gap: 42px` ;)

File: packages/suite/src/views/wallet/transactions/TradeBox/TradeBoxPrices.tsx
Comment (dahaca) [inline packages/suite/src/views/wallet/transactions/TradeBox/TradeBoxPrices.tsx]: same as `Ticker` pretty much, so some util could be made

**Review by dahaca — COMMENTED** [review]
Is there a reason why the design differs from the one in https://github.com/trezor/trezor-suite/issues/9359? 🤔

**Comment by FreeWall** [comment]
> Is there a reason why the design differs from the one in #9359? 🤔

Design was changed in task since then, #9359 updated

File: packages/suite/src/components/suite/OverflowItems.tsx
Comment (FreeWall) [reply packages/suite/src/components/suite/OverflowItems.tsx]: not really, in AppNavigation all buttons are hidden if one of them doesn't fit, here we want to gradually hide buttons one by one

File: packages/suite/src/components/suite/OverflowItems.tsx
Comment (FreeWall) [reply packages/suite/src/components/suite/OverflowItems.tsx:20]: I wanted to make a more generic component to accept any elements, not just buttons

File: suite-common/wallet-core/src/fiat-rates/fiatRatesMiddleware.ts
Comment (FreeWall) [reply suite-common/wallet-core/src/fiat-rates/fiatRatesMiddleware.ts]: In mobile viewports or mobile app? I'm not sure what you mean

File: suite-common/wallet-core/src/fiat-rates/fiatRatesThunks.ts
Comment (FreeWall) [reply suite-common/wallet-core/src/fiat-rates/fiatRatesThunks.ts]: Yes good point, <s>I can make this more generic with weeks argument.</s>

Btw, are you talking about the existing `updateLastWeekFiatRatesThunk` too? Because this does something a little bit different and I didn't want to break this feature anywhere..
Thing is, `updateLastWeekFiatRatesThunk` gets the data for 1 day only (7 days ago) for all fiat currencies, so the description of this function is wrong (it doesn't get 168 data points, only 1).
My first thought initially was to have _updateLast**Two**WeekFiatRatesThunk_ only and calculate `lastWeek` object from this data (to eliminate duplicate API calls), but `lastWeek` has rates for all fiat currencies and for this new TradeBox component we need rates for 14 days in past, which might be a lot of data for all fiat currencies (10+ KB), so instead of fetching the data for all currencies I made this `localCurrency` argument which fetch the data only for a single fiat currency which is a lot more smaller in size.
I'm not sure about this really, what do you think?

**Comment by matejkriz** [comment]
Please wait for product review from @Hermez-cz before merge 🙏  Personally, I miss being able to hide it somehow.

**Comment by matejkriz** [comment]
@FreeWall check this PR https://github.com/trezor/trezor-suite/pull/9292 please, I guess there are conflicts.

**Comment by FreeWall** [comment]
> @FreeWall check this PR #9292 please, I guess there are conflicts.

@matejkriz Is it going to be merged in the near future? Or should I temporarily move my code of fiat rates elsewhere for now?

File: packages/suite/src/components/suite/OverflowItems.tsx
Comment (FreeWall) [reply packages/suite/src/components/suite/OverflowItems.tsx]: Yes, it shows all items and then gradually hide one by one and checking the content width with bounding width

File: packages/suite/src/views/wallet/transactions/TradeBox/TradeBoxMenu.tsx
Comment (FreeWall) [reply packages/suite/src/views/wallet/transactions/TradeBox/TradeBoxMenu.tsx]: isn't a part of the approved design

File: packages/suite/src/views/wallet/transactions/TradeBox/TradeBoxPrices.tsx
Comment (FreeWall) [reply packages/suite/src/views/wallet/transactions/TradeBox/TradeBoxPrices.tsx]: what you mean, util to compare two numbers? :D

File: suite-common/wallet-core/src/fiat-rates/fiatRatesMiddleware.ts
Comment (dahaca) [reply suite-common/wallet-core/src/fiat-rates/fiatRatesMiddleware.ts]: I meant the mobile app. Actually, my bad, the location of this file confused me, it being `suite-common`, there is a different `fiatRatesMiddleware` in `suite-native` and this one is only used in suite

File: packages/suite/src/views/wallet/transactions/TradeBox/TradeBoxPrices.tsx
Comment (dahaca) [reply packages/suite/src/views/wallet/transactions/TradeBox/TradeBoxPrices.tsx]: Yep, to get `rateGoingUp` based on current rates and rated from the past.

File: suite-common/wallet-core/src/fiat-rates/fiatRatesThunks.ts
Comment (dahaca) [reply suite-common/wallet-core/src/fiat-rates/fiatRatesThunks.ts]: I am and yeah, I've noticed that it gets 1 data point and wasn't sure if it's just bad naming or a bug. Well, is there a reason to get data for all fiat currencies though? 🤔 It seems like `localCurrency` is always used to access the rates, and those rates can be refetched whenever it changes. And then, as you say, everything can be fetched once to be used in both places :)

File: packages/suite/src/components/suite/OverflowItems.tsx
Comment (dahaca) [reply packages/suite/src/components/suite/OverflowItems.tsx]: `OverflowItems` seems like improved behavior to me + we would avoid logic/code duplication. Looks like an improvement to me, wdyt? @matejkriz @Hermez-cz

File: packages/suite/src/components/suite/OverflowItems.tsx
Comment (dahaca) [reply packages/suite/src/components/suite/OverflowItems.tsx]: then how about instead of `repartition(false)` just use this directly? 
 
```
setState({
    overflowItems: [],
    visibleItems: props.items,
})
```
                
Or have another function:

```
const reset = () => setState({
    overflowItems: [],
    visibleItems: props.items,
})
```

File: packages/suite/src/components/suite/OverflowItems.tsx
Comment (dahaca) [reply packages/suite/src/components/suite/OverflowItems.tsx:20]: I don't think we will have any use cases for other than buttons anytime soon though 🤔

File: packages/suite/src/views/wallet/transactions/TradeBox/TradeBoxMenu.tsx
Comment (dahaca) [reply packages/suite/src/views/wallet/transactions/TradeBox/TradeBoxMenu.tsx]: There is no approved mobile design though, is there? :D And it's cleaner without that free space on the right, it's a minor but nice change IMO

File: suite-common/wallet-core/src/fiat-rates/fiatRatesThunks.ts
Comment (FreeWall) [reply suite-common/wallet-core/src/fiat-rates/fiatRatesThunks.ts]: Yea it should be refetched in case of localCurrency changed, I wasn't sure really why it was designed this way and if I'm not missing something..
I'll try to merge them into one then

File: packages/suite/src/views/wallet/transactions/TradeBox/TradeBoxMenu.tsx
Comment (FreeWall) [reply packages/suite/src/views/wallet/transactions/TradeBox/TradeBoxMenu.tsx]: It is, I follow Figma design from Invity task, which is missing in Issue, I can update it with more screenshots

File: packages/suite/src/components/suite/OverflowItems.tsx
Comment (FreeWall) [reply packages/suite/src/components/suite/OverflowItems.tsx:20]: Isn't that a bit of shame to not have a generic one and to make additional dependency which can't be changed from outside? seems like an antipattern to me when the job is already done.

File: suite-common/wallet-core/src/fiat-rates/fiatRatesThunks.ts
Comment (FreeWall) [reply suite-common/wallet-core/src/fiat-rates/fiatRatesThunks.ts]: https://github.com/trezor/trezor-suite/pull/9722/commits/1f640b276ac910729494a8aeae9441d96ecf2dbd Here I combined them into a single API request and eliminated some of the duplicities, in state there is still lastWeek and lastTwoWeeks though, what do you think?

File: packages/suite/src/components/suite/OverflowItems.tsx
Comment (dahaca) [reply packages/suite/src/components/suite/OverflowItems.tsx:20]: True, engineering-wise it's not a bad solution at all! All I am saying is that I don't think we are going to have any other cases other than those buttons in the `TradeBox` and `AppNavigation` + Suite's design is about to change. Sometimes it's better to have a less generic, simpler solution to avoid having unnecessary complications later down the line. Since it's already done – leave it there :)

File: packages/suite/src/views/wallet/transactions/TradeBox/TradeBoxMenu.tsx
Comment (dahaca) [reply packages/suite/src/views/wallet/transactions/TradeBox/TradeBoxMenu.tsx]: I see now! Ok then 😅

File: suite-common/wallet-core/src/fiat-rates/fiatRatesThunks.ts
Comment (dahaca) [inline suite-common/wallet-core/src/fiat-rates/fiatRatesThunks.ts]: why is this changed from `&&` to `||`?

File: suite-common/wallet-core/src/fiat-rates/fiatRatesThunks.ts
Comment (dahaca) [inline suite-common/wallet-core/src/fiat-rates/fiatRatesThunks.ts]: To avoid duplication even further, how about:

* fetching for the last 2 weeks and storing that like so:
```
 fiatRatesActions.updateLastWeeksFiatRates({
    symbol: ticker.symbol,
    tickers: results.tickers,
    ts: new Date().getTime(),
})
```

* adding a selector `selectRatedFromAWeekAgo()` which would return `coin.lastTwoWeekRates.tickers[6]`

Then you don't need to have `lastWeek` at all and duplicate `symbol` and `ts`, as they always will be the same :)

**Review by dahaca — CHANGES_REQUESTED** [review]
I am leaving those comments just for you but @matejkriz was right – [this PR](https://github.com/trezor/trezor-suite/pull/9292) changes a lot and you'd have to wait for it to be merged and rebase then :( 

For example, it [moves](https://github.com/trezor/trezor-suite/blob/f09648a46199d22b4532a0f3c68d08cdbe7f3bb9/suite-common/wallet-core/src/fiat-rates/fiatRatesThunks.ts#L84) the approach of **suite-mobile** to **common**, which would affect how you should implement adding `twoWeekRates`. In fact things I wrote in the new comment regarding fetching 2 weeks and then using a selector to pick a rate from a week ago would likely still apply. 

My bad for asking you to do more reducer-related changes now, shouldn't have done that until rates refactor pr is merged 😓

**Comment by Hermez-cz** [comment]
Can we please make it foldable? Ondra can suggest design solution if needed, but a simple chevron will do the job.

File: packages/suite/src/components/suite/OverflowItems.tsx
Comment (dahaca) [inline packages/suite/src/components/suite/OverflowItems.tsx]: can this all be in on `useEffect` or `useUpdateEffect`?

File: packages/suite/src/views/wallet/transactions/TradeBox/TradeBox.tsx
Comment (FreeWall) [reply packages/suite/src/views/wallet/transactions/TradeBox/TradeBox.tsx]: https://github.com/trezor/trezor-suite/pull/9722/commits/3d40bee59bf097b378812ae377e3585cf52e7179 fixed

File: packages/suite/src/components/suite/OverflowItems.tsx
Comment (dahaca) [inline packages/suite/src/components/suite/OverflowItems.tsx]: nit: `false` is not really needed since it's falsy when `undefined`

File: packages/suite/src/components/suite/OverflowItems.tsx
Comment (dahaca) [inline packages/suite/src/components/suite/OverflowItems.tsx]: would be nice to keep component state and selectors above callbacks and effects unless not possible. It helps readability. It's not strictly enforced but everyone more or less is following [the same order](https://www.notion.so/satoshilabs/Components-9159b639d21b43ef8b66be150854b6b6?pvs=4#308aa7db519e4f4eaf70312f60ce030a).

File: packages/suite-analytics/src/constants.ts
Comment (FreeWall) [reply packages/suite-analytics/src/constants.ts:39]: We might track other components in tradebox too in the future, tradebox/button seems good to me

File: packages/suite/src/views/wallet/transactions/TradeBox/TradeBox.tsx
Comment (FreeWall) [reply packages/suite/src/views/wallet/transactions/TradeBox/TradeBox.tsx:108]: Maybe in the next iteration of this task? I don't want to do any futher changes until #9292 is done

**Review by dahaca — COMMENTED** [review]
Other than mentioned nitpicks, the code is OK. Please use `OverflowItems` in the `AppNavigation` in the future to make the logic consistent.

Not approving yet, waiting for the green light from @Hermez-cz

File: packages/suite/src/components/suite/OverflowItems.tsx
Comment (FreeWall) [reply packages/suite/src/components/suite/OverflowItems.tsx]: removed some useless hooks and renamed functions to be more understandable

File: packages/suite/src/views/wallet/transactions/TradeBox/TradeBoxPrices.tsx
Comment (FreeWall) [reply packages/suite/src/views/wallet/transactions/TradeBox/TradeBoxPrices.tsx]: Maybe in the next iteration of this task? I don't want to do any futher changes until https://github.com/trezor/trezor-suite/pull/9292 is done

File: packages/suite/src/components/suite/OverflowItems.tsx
Comment (FreeWall) [reply packages/suite/src/components/suite/OverflowItems.tsx]: https://github.com/trezor/trezor-suite/pull/9722/commits/9dacbbb1e944ac9b4042d3097509aff4a2207a08 indeed

File: packages/suite/src/components/suite/OverflowItems.tsx
Comment (FreeWall) [reply packages/suite/src/components/suite/OverflowItems.tsx]: https://github.com/trezor/trezor-suite/pull/9722/commits/9dacbbb1e944ac9b4042d3097509aff4a2207a08 fixed

File: packages/suite/src/components/suite/OverflowItems.tsx
Comment (FreeWall) [reply packages/suite/src/components/suite/OverflowItems.tsx]: https://github.com/trezor/trezor-suite/pull/9722/commits/9dacbbb1e944ac9b4042d3097509aff4a2207a08 fixed

**Review by dahaca — APPROVED** [review]

**Comment by Hermez-cz** [comment]
> Other than mentioned nitpicks, the code is OK. Please use `OverflowItems` in the `AppNavigation` in the future to make the logic consistent.
> 
> Not approving yet, waiting for the green light from @Hermez-cz

Thank you, approved including today's changes discussed with @marek-andrsk

**Pull request merged**
